src/main/java/de/tum/in/www1/artemis/web/rest/GenericSubmissionResource.java

```diff
@@ -0,0 +1,241 @@
+        }
+        E exercise;
+        if (exerciseType.isInstance(participation.getExercise())) {
+            exercise = exerciseType.cast(participation.getExercise());
+            if (exercise == null) {
+                return ResponseEntity.badRequest()
+                        .headers(HeaderUtil.createFailureAlert(applicationName, true, "exercise", "exerciseEmpty", "The exercise belonging to the participation is null."))
+                        .body(null);
+            }
+
+            // make sure sensitive information are not sent to the client
+            exercise.filterSensitiveInformation();
+        }
+        else {
+            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, "fileUploadExercise", "wrongExerciseType",
+                    "The exercise of the participation is not a file upload exercise.")).body(null);
+        }
+
+        // Students can only see their own file uploads (to prevent cheating). TAs, instructors and admins can see all file uploads.
+        if (!(authCheckService.isOwnerOfParticipation(participation) || authCheckService.isAtLeastTeachingAssistantForExercise(exercise))) {
+            return forbidden();
+        }
+
+        Optional<T> optionalSubmission = participation.findLatestSubmissionOfType(submissionType);
+        T submission;
+        if (optionalSubmission.isEmpty()) {
+            // this should never happen as the submission is initialized along with the participation when the exercise is started
+            submission = newSubmission;
+            submission.setParticipation(participation);
+        }
+        else {
+            // only try to get and set the file upload if the fileUploadSubmission existed before
+            submission = optionalSubmission.get();
+        }
+
+        // make sure only the latest submission and latest result is sent to the client
+        participation.setSubmissions(null);
+        participation.setResults(null);
+
+        if (submission.getResult() != null) {
+            // do not send the result to the client if the assessment is not finished
+            if (submission.getResult().getCompletionDate() == null || submission.getResult().getAssessor() == null) {
+                submission.setResult(null);
+            }
+            if (!authCheckService.isAtLeastTeachingAssistantForExercise(exercise)) {
+                submission.getResult().setAssessor(null);
+            }
+        }
+        return ResponseEntity.ok(submission);
+    }
+
+    /**
+     * Removes sensitive information (e.g. example solution of the exercise) from the submission based on the role of the current user. This should be called before sending a
+     * submission to the client.
+     * ***IMPORTANT***: Do not call this method from a transactional context as this would remove the sensitive information also from the entities in the
```

Comment:
I think, we can check if there is an open transaction using `TransactionSynchronizationManager.isActualTransactionActive()` and maybe throw an exception in these cases. Since this is so critical in this case, this might be a good idea.

Reply:
Very good idea. I didn't know about this method.